### PR TITLE
Add support for modifying what cluster role bindings are ignored

### PR DESCRIPTION
**Pull request by mprahl**
Before this commit, any cluster role binding that started with "system"
was ignored. The intention was to ignore any default cluster role
bindings generated by Kubernetes/OpenShift. This was too liberal since
it also ignored cluster role bindings named things like
"system-administrators". The default is thus changed to "system:".

Since there is nothing restricting users of creating cluster role bindings
that start with "system:", a new field on the IamPolicy kind was added
called "spec.ignoreClusterRoleBindings". This is an array of regex
values that can be used to ignore cluster role bindings that the
user desires. This means they can have a stricter regex or disable it
entirely with a regex such as `.^`.

Resolves: https://github.com/open-cluster-management/backlog/issues/14874

### CODE REVIEW

**Comment by ycao56** [comment]
I am not familiar with iam policy controller code. So adding @ckandag and @gparvin to take a look.

Does OCP have any default role that would be ignored previously but not anymore after this PR is merged? If yes, we may also need to update our test cases in grc-ui and policy-framework.

**Comment by mprahl** [comment]
> I am not familiar with iam policy controller code. So adding @ckandag and @gparvin to take a look.
> 
> Does OCP have any default role that would be ignored previously but not anymore after this PR is merged? If yes, we may also need to update our test cases in grc-ui and policy-framework.

I think the stricter match shouldn't match a normal use-case since these are the default cluster role bindings with the name "system" and not "system:" in them on a cluster with RHACM installed:

```bash
❯ oc get clusterrolebinding -o=jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}' | grep system | grep -v "system:"
system-bootstrap-node-bootstrapper
system-bootstrap-node-renewal
```

Those bindings reference these roles:
* system:node-bootstrapper
* system:certificates.k8s.io:certificatesigningrequests:selfnodeclient

I don't think users would have an IAM policy for those roles.

**Comment by ycao56** [comment]
Agree. I am not worried about this change. Just want to see if it would break canary.

File: deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml [hunk 1]
```diff
@@ -32,6 +32,15 @@ spec:
           spec:
             description: IamPolicySpec defines the desired state of IamPolicy
             properties:
+              ignoreClusterRoleBindings:
```
Comment (JustinKuli) [inline deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml:35]: I noticed that we have 2 copies of this CRD - one here at `deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml` and another at `deploy/crds/policy.open-cluster-management.io_iampolicies_crd.yaml`.

The `deploy` make target runs `kubectl apply -f deploy/crds/ ...` so I don't think that will pick up this change. But the `install-crds` make target does apply this (changed) crd yaml...

tl;dr, can we remove whatever the extra CRD YAML is, and ensure the make targets still work. I think it's possible to address this in another PR if you'd prefer.

File: pkg/apis/policy/v1/iampolicy_types.go [hunk 2]
```diff
@@ -45,6 +45,8 @@ type Target struct {
 
 // IamPolicySpec defines the desired state of IamPolicy
 type IamPolicySpec struct {
+	// This defaults to ignoring all cluster role binding names that start with "system:"
```
Comment (JustinKuli) [inline pkg/apis/policy/v1/iampolicy_types.go:48]: I think this comment is supposed to match the `description` in the CRD... I couldn't get the old operator-sdk version installed and generating this CRD anyway, so it might not matter now. Just something to be aware of when we do upgrade that. Operator-sdk v1+ will take the comments in this struct and populate the relevant fields in the CRD.

**Review by JustinKuli — COMMENTED** [review]
Looks good. Some CRD upkeep concerns, but they aren't strictly necessary now.

File: deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml (hunk 1 above)
Comment (ycao56) [reply deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml:35]: This reminds me that any CRD change on managed cluster side needs to make into https://github.com/open-cluster-management/klusterlet-addon-operator/blob/main/helm-charts

File: pkg/apis/policy/v1/iampolicy_types.go (hunk 2 above)
Comment (mprahl) [reply pkg/apis/policy/v1/iampolicy_types.go:48]: Got it. I didn't think we used the operator-sdk to generate these manifests based on the directory structure of `deploy`. I'll update it though.

Comment (mprahl) [reply pkg/apis/policy/v1/iampolicy_types.go]: @JustinKuli you are right about that. I updated the comment above the struct field.

operator-sdk manifest and code generation is currently broken in this repo. The manifest generation fails with this:
```
iam-policy-controller on  allowlist-regex [?] via 🐹 v1.16.8 ❯ ./operator-sdk generate crds     
[Deprecation Notice] Operator SDK has a new CLI and project layout that is aligned with Kubebuilder.
See `operator-sdk init -h` and the following doc on how to scaffold a new project:
https://v0-19-x.sdk.operatorframework.io/docs/golang/quickstart/
To migrate existing projects to the new layout see:
https://sdk.operatorframework.io/docs/building-operators/golang/project_migration_guide/

INFO[0000] Running CRD generator.                       
/home/mprahl/git/iam-policy-controller/pkg/apis/policy/v1/iampolicy_types.go:69:31: map values must be a named type, not *ast.MapType
FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
```
It's failing on this:
https://github.com/open-cluster-management/iam-policy-controller/blob/main/pkg/apis/policy/v1/iampolicy_types.go#L65

Which was introduced over a year ago in https://github.com/open-cluster-management/iam-policy-controller/commit/e06184dca540757c2b10f12e2f72d744c4f4b697.

Comment (JustinKuli) [reply pkg/apis/policy/v1/iampolicy_types.go]: Good to know. And I agree - we can just fix this later when updating the operator-sdk.

File: deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml (hunk 1 above)
Comment (mprahl) [reply deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml:35]: After discussing this with @JustinKuli, we decided to also include the change in the v1beta1 CRD rather than deleting it.

Comment (mprahl) [reply deploy/crds/v1/policy.open-cluster-management.io_iampolicies.yaml:35]: @ycao56 I'll make the change there as well if this PR gets approved.

**Review by gparvin — APPROVED** [review]
This looks good.  Thanks
